core-models/src/main/scala/com/mesosphere/usi/core/models/PodSpec.scala

```diff
@@ -12,4 +16,62 @@ package com.mesosphere.usi.core.models
   * @param goal target goal of this pod. See [[Goal]] for more details
   * @param runSpec WIP the thing to run, and resource requirements, etc.
   */
-case class PodSpec(id: PodId, goal: Goal, runSpec: RunSpec)
+trait PodSpec {
+  def id: PodId
+  def goal: Goal
+  def runSpec: RunSpec
+}
+
+object PodSpec {
+
+  type ValidationError = String
+
+  /**
+    * Verifies that every value in range is requested only once. Requesting same value multiple times would yield pod that will never be scheduled.
+    * E.g. requesting two ports 80 on one machine is just not possible to satisfy
+    * @param runSpec runSpec we are validating
+    * @return true if provided range requirements are valid
+    */
+  private def validateStaticRangeRequirementsUnique(runSpec: RunSpec): Seq[ValidationError] = {
+    val staticPorts = runSpec.resourceRequirements.collect {
+      case RangeRequirement(requestedValues, _, _) => requestedValues
+    }.flatten.collect { case ExactValue(value) => value }
+
+    if (staticPorts.distinct.length != staticPorts.length) {
+      Seq(s"Every value inside RangeResource can be requested only once. Requirement: ${staticPorts.mkString(",")}")
+    } else {
+      Seq.empty
+    }
+  }
+
+  private def validateScalarRequirements(resourceRequirements: Seq[ResourceRequirement]): Seq[ValidationError] = {
+    val invalidScalar = resourceRequirements.collect {
+      case s: ScalarRequirement => s
+    }.filter(s => s.amount < 0)
+    if (invalidScalar.nonEmpty) {
+      Seq(s"Scalar values cannot be smaller than 0. Invalid requirements: ${invalidScalar
+        .map(s => s"${s.resourceType}:${s.amount}")
+        .mkString(",")}")
+    } else {
+      Seq.empty
+    }
+  }
+
+  def apply(id: PodId, goal: Goal, runSpec: RunSpec): Either[Seq[ValidationError], PodSpec] = {
+    val uniqueRangeRequirements = validateStaticRangeRequirementsUnique(runSpec)
+    val scalarRequirements = validateScalarRequirements(runSpec.resourceRequirements)
+
+    validationErrorOrResult(uniqueRangeRequirements, scalarRequirements)(() => PodSpecImpl(id, goal, runSpec))
+  }
+
+  private def validationErrorOrResult(validationResults: Seq[ValidationError]*)(createFunc: () => PodSpec): Either[Seq[ValidationError], PodSpec] = {
```

Comment:
So we are building our own validation library? 😉

Reply:
no we implemented 5 lines of code instead of pulling in generic utility library exactly because of these 5 lines of code.

Reply:
I’m just afraid that we’ll put more stuff on top instead of following some common pattern.

Reply:
then we should go for a proper validation library at that point. My argument here is that we don't need those types from Scalactic. Maybe at one point full featured validation library would be a way to go although I kind of like the simplicity and the conventions outlined here...

Reply:
My vote is a firm no for using a validation library.